### PR TITLE
dialects: Add a stencil.ApplyOp to scf.ParallelOp lowering.

### PR DESCRIPTION
**Pull request by PapyChacal**

### CODE REVIEW

**Review by webmiche — APPROVED** [review]
LGTM

**Review by superlopuh — APPROVED** [review]

**Review by georgebisbas — COMMENTED** [review]
Looks cool, are there cases that may cause OOB accesses? Are they tested?

File: xdsl/dialects/experimental/stencil.py [hunk 1]
```diff
@@ -131,6 +131,13 @@ def verify(self) -> None:
                 f"Expected 3 indexes for stencil.index, got {len(self.array.data)}."
             )
 
+    @staticmethod
+    def sizeFromBounds(lb: IndexAttr, ub: IndexAttr) -> Sequence[int]:
```
Comment (AntonLydike) [inline xdsl/dialects/experimental/stencil.py:135]: I think the naming convention for python method and function names is `snake_case`. So this should be `size_from_bounds`.

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py [hunk 2]
```diff
@@ -9,30 +9,27 @@
...
-from xdsl.dialects.experimental.stencil import CastOp, FieldType, IndexAttr, LoadOp
+from xdsl.dialects.experimental.stencil import ApplyOp, CastOp, FieldType, IndexAttr, LoadOp, TempType
 
 _TypeElement = TypeVar("_TypeElement", bound=Attribute)
 
 
 def GetMemRefFromField(
-        inputFieldType: FieldType[_TypeElement]) -> MemRefType[_TypeElement]:
-    dims = [i.value.data for i in inputFieldType.shape.data]
+    inputType: FieldType[_TypeElement] | TempType[_TypeElement]
```
Comment (AntonLydike) [inline xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:20]: I think most fields should also be `snake_case`. Although we are quite inconsistent with that. Although most cases in our code where we do this are (hopefully) for external api compatibility (e.g. MLIR might not use snake case for attribute names)

**Review by AntonLydike — COMMENTED** [review]
Can't comment on the rewrites themselves as I am not really familiar with stencil (yet). Just found some minor issues with formatting conventions. Otherwise LGTM

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:20]: Oh, well this is similar here, they're from open earth.
I can switch to snake. Is there a way to check that automatically?

File: xdsl/dialects/experimental/stencil.py (hunk 1 above)
Comment (PapyChacal) [reply xdsl/dialects/experimental/stencil.py:135]: Will do

File: xdsl/transforms/experimental/ConvertStencilToLLMLIR.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/transforms/experimental/ConvertStencilToLLMLIR.py:20]: Nah, got confused, just corrected!

**Comment by PapyChacal** [comment]
Let's 

> Looks cool, are there cases that may cause OOB accesses? Are they tested?

Not yet, but that will be pretty easy to check for, given static bounds!

**Pull request merged**